<a href="https://colab.research.google.com/github/lucca-fsc/ChallengeDataScience/blob/main/Challenge_Data_Science.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Challenge DataScience 

## Para este Challenge, fomos contratados pela empresa de Telecomunicações Alura Voz para realizar uma análise dos dados e buscar entender e reduzir a taxa de evasão dos clientes, também conhecido como *Churn*. 

#### Dicionário de dados

* `customerID`: número de identificação único de cada cliente
* `Churn`: se o cliente deixou ou não a empresa 
* `gender`: gênero (masculino e feminino) 
* `SeniorCitizen`: informação sobre um cliente ter ou não idade igual ou maior que 65 anos 
* `Partner`:  se o cliente possui ou não um parceiro ou parceira
* `Dependents`: se o cliente possui ou não dependentes
* `tenure`:  meses de contrato do cliente
* `PhoneService`: assinatura de serviço telefônico 
* `MultipleLines`: assisnatura de mais de uma linha de telefone 
* `InternetService`: assinatura de um provedor internet 
* `OnlineSecurity`: assinatura adicional de segurança online 
* `OnlineBackup`: assinatura adicional de backup online 
* `DeviceProtection`: assinatura adicional de proteção no dispositivo 
* `TechSupport`: assinatura adicional de suporte técnico, menos tempo de espera
* `StreamingTV`: assinatura de TV a cabo 
* `StreamingMovies`: assinatura de streaming de filmes 
* `Contract`: tipo de contrato
* `PaperlessBilling`: se o cliente prefere receber online a fatura
* `PaymentMethod`: forma de pagamento
* `Charges.Monthly`: total de todos os serviços do cliente por mês
* `Charges.Total`: total gasto pelo cliente

## Primeira Semana 

### As Tarefas para a primeira semana do Challenge consistem em abrir a base de dados, e realizar algumas primeiras análise dos dados

#### Importando as bibliotecas e Fazendo a primeira leitura da base de dados

In [1]:
import pandas as pd
import numpy as np

##### Dados foram trazidos de uma API em formato *json*, é necessário realizar um tratamento especial para conseguir montar um dataframe com todas as informações necessárias.


In [2]:
dados = pd.read_json(
    path_or_buf= 'Telco-Customer-Churn.json',
    orient = 'columns'
)

dados

customerID Churn                                           customer  \
0     0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1     0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2     0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3     0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4     0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   
...          ...   ...                                                ...   
7262  9987-LUTYD    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
7263  9992-RRAMN   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7264  9992-UJOEL    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7265  9993-LHIEB    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
7266  9995-HOTOH    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   

                                                  phone  \
0        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1       {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4        {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
...                                                 ...   
7262     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7263    {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
7264     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7265     {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
7266  {'PhoneService': 'No', 'MultipleLines': 'No ph...   

                                               internet  \
0     {'InternetService': 'DSL', 'OnlineSecurity': '...   
1     {'InternetService': 'DSL', 'OnlineSecurity': '...   
2     {'InternetService': 'Fiber optic', 'OnlineSecu...   
3     {'InternetService': 'Fiber optic', 'OnlineSecu...   
4     {'InternetService': 'Fiber optic', 'OnlineSecu...   
...                                                 ...   
7262  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7263  {'InternetService': 'Fiber optic', 'OnlineSecu...   
7264  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7265  {'InternetService': 'DSL', 'OnlineSecurity': '...   
7266  {'InternetService': 'DSL', 'OnlineSecurity': '...   

                                                account  
0     {'Contract': 'One year', 'PaperlessBilling': '...  
1     {'Contract': 'Month-to-month', 'PaperlessBilli...  
2     {'Contract': 'Month-to-month', 'PaperlessBilli...  
3     {'Contract': 'Month-to-month', 'PaperlessBilli...  
4     {'Contract': 'Month-to-month', 'PaperlessBilli...  
...                                                 ...  
7262  {'Contract': 'One year', 'PaperlessBilling': '...  
7263  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7264  {'Contract': 'Month-to-month', 'PaperlessBilli...  
7265  {'Contract': 'Two year', 'PaperlessBilling': '...  
7266  {'Contract': 'Two year', 'PaperlessBilling': '...  

[7267 rows x 6 columns]

##### Foram então normalizadas todas as colunas dos datasets para ter acesso a todo o seu conteúdo.

In [3]:
dados_normal_customer = pd.json_normalize(data = dados.customer, sep = '_')
dados_normal_customer

gender  SeniorCitizen Partner Dependents  tenure
0     Female              0     Yes        Yes       9
1       Male              0      No         No       9
2       Male              0      No         No       4
3       Male              1     Yes         No      13
4     Female              1     Yes         No       3
...      ...            ...     ...        ...     ...
7262  Female              0      No         No      13
7263    Male              0     Yes         No      22
7264    Male              0      No         No       2
7265    Male              0     Yes        Yes      67
7266    Male              0     Yes        Yes      63

[7267 rows x 5 columns]

In [4]:
dados_normal_phone = pd.json_normalize(data = dados.phone, sep = '_')
dados_normal_phone

PhoneService     MultipleLines
0             Yes                No
1             Yes               Yes
2             Yes                No
3             Yes                No
4             Yes                No
...           ...               ...
7262          Yes                No
7263          Yes               Yes
7264          Yes                No
7265          Yes                No
7266           No  No phone service

[7267 rows x 2 columns]

In [5]:
dados_normal_internet = pd.json_normalize(data = dados.internet, sep = '_')
dados_normal_internet

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0                DSL             No          Yes               No         Yes   
1                DSL             No           No               No          No   
2        Fiber optic             No           No              Yes          No   
3        Fiber optic             No          Yes              Yes          No   
4        Fiber optic             No           No               No         Yes   
...              ...            ...          ...              ...         ...   
7262             DSL            Yes           No               No         Yes   
7263     Fiber optic             No           No               No          No   
7264             DSL             No          Yes               No          No   
7265             DSL            Yes           No              Yes         Yes   
7266             DSL            Yes          Yes              Yes          No   

     StreamingTV StreamingMovies  
0            Yes              No  
1             No             Yes  
2             No              No  
3            Yes             Yes  
4            Yes              No  
...          ...             ...  
7262          No              No  
7263          No             Yes  
7264          No              No  
7265          No             Yes  
7266         Yes             Yes  

[7267 rows x 7 columns]

In [6]:
dados_normal_account = pd.json_normalize(data = dados.account, sep = '_')
dados_normal_account

Contract PaperlessBilling     PaymentMethod  Charges_Monthly  \
0           One year              Yes      Mailed check            65.60   
1     Month-to-month               No      Mailed check            59.90   
2     Month-to-month              Yes  Electronic check            73.90   
3     Month-to-month              Yes  Electronic check            98.00   
4     Month-to-month              Yes      Mailed check            83.90   
...              ...              ...               ...              ...   
7262        One year               No      Mailed check            55.15   
7263  Month-to-month              Yes  Electronic check            85.10   
7264  Month-to-month              Yes      Mailed check            50.30   
7265        Two year               No      Mailed check            67.85   
7266        Two year               No  Electronic check            59.00   

     Charges_Total  
0            593.3  
1            542.4  
2           280.85  
3          1237.85  
4            267.4  
...            ...  
7262         742.9  
7263        1873.7  
7264         92.75  
7265       4627.65  
7266        3707.6  

[7267 rows x 5 columns]

In [7]:
dados_classification = dados[['customerID', 'Churn']]

##### Após o Final das normalizações, basta fazer a junção de todos os dataframes que foram criados no momento da normalização.

In [8]:
dados_total = pd.merge(
    left = dados_normal_customer,
    right = dados_normal_phone,
    left_index = True,
    right_index = True
)

In [9]:
dados_total = pd.merge(
    left = dados_total,
    right = dados_normal_internet,
    left_index = True,
    right_index = True
)
dados_total

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes        Yes       9          Yes   
1       Male              0      No         No       9          Yes   
2       Male              0      No         No       4          Yes   
3       Male              1     Yes         No      13          Yes   
4     Female              1     Yes         No       3          Yes   
...      ...            ...     ...        ...     ...          ...   
7262  Female              0      No         No      13          Yes   
7263    Male              0     Yes         No      22          Yes   
7264    Male              0      No         No       2          Yes   
7265    Male              0     Yes        Yes      67          Yes   
7266    Male              0     Yes        Yes      63           No   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0                   No             DSL             No          Yes   
1                  Yes             DSL             No           No   
2                   No     Fiber optic             No           No   
3                   No     Fiber optic             No          Yes   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7262                No             DSL            Yes           No   
7263               Yes     Fiber optic             No           No   
7264                No             DSL             No          Yes   
7265                No             DSL            Yes           No   
7266  No phone service             DSL            Yes          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies  
0                  No         Yes         Yes              No  
1                  No          No          No             Yes  
2                 Yes          No          No              No  
3                 Yes          No         Yes             Yes  
4                  No         Yes         Yes              No  
...               ...         ...         ...             ...  
7262               No         Yes          No              No  
7263               No          No          No             Yes  
7264               No          No          No              No  
7265              Yes         Yes          No             Yes  
7266              Yes          No         Yes             Yes  

[7267 rows x 14 columns]

In [10]:
dados_total = pd.merge(
    left = dados_total,
    right = dados_normal_account,
    left_index = True,
    right_index = True
)
dados_total

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes        Yes       9          Yes   
1       Male              0      No         No       9          Yes   
2       Male              0      No         No       4          Yes   
3       Male              1     Yes         No      13          Yes   
4     Female              1     Yes         No       3          Yes   
...      ...            ...     ...        ...     ...          ...   
7262  Female              0      No         No      13          Yes   
7263    Male              0     Yes         No      22          Yes   
7264    Male              0      No         No       2          Yes   
7265    Male              0     Yes        Yes      67          Yes   
7266    Male              0     Yes        Yes      63           No   

         MultipleLines InternetService OnlineSecurity OnlineBackup  \
0                   No             DSL             No          Yes   
1                  Yes             DSL             No           No   
2                   No     Fiber optic             No           No   
3                   No     Fiber optic             No          Yes   
4                   No     Fiber optic             No           No   
...                ...             ...            ...          ...   
7262                No             DSL            Yes           No   
7263               Yes     Fiber optic             No           No   
7264                No             DSL             No          Yes   
7265                No             DSL            Yes           No   
7266  No phone service             DSL            Yes          Yes   

     DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0                  No         Yes         Yes              No        One year   
1                  No          No          No             Yes  Month-to-month   
2                 Yes          No          No              No  Month-to-month   
3                 Yes          No         Yes             Yes  Month-to-month   
4                  No         Yes         Yes              No  Month-to-month   
...               ...         ...         ...             ...             ...   
7262               No         Yes          No              No        One year   
7263               No          No          No             Yes  Month-to-month   
7264               No          No          No              No  Month-to-month   
7265              Yes         Yes          No             Yes        Two year   
7266              Yes          No         Yes             Yes        Two year   

     PaperlessBilling     PaymentMethod  Charges_Monthly Charges_Total  
0                 Yes      Mailed check            65.60         593.3  
1                  No      Mailed check            59.90         542.4  
2                 Yes  Electronic check            73.90        280.85  
3                 Yes  Electronic check            98.00       1237.85  
4                 Yes      Mailed check            83.90         267.4  
...               ...               ...              ...           ...  
7262               No      Mailed check            55.15         742.9  
7263              Yes  Electronic check            85.10        1873.7  
7264              Yes      Mailed check            50.30         92.75  
7265               No      Mailed check            67.85       4627.65  
7266               No  Electronic check            59.00        3707.6  

[7267 rows x 19 columns]

##### No final obtivemos o DataFrame (dados_total) que contém todos os dados recuperados em um único dataset, facilitando o estudo e acesso às informações.

In [11]:
dados_total = pd.merge(
    left = dados_total,
    right = dados_classification,
    left_index = True,
    right_index = True
)
dados_total

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes        Yes       9          Yes   
1       Male              0      No         No       9          Yes   
2       Male              0      No         No       4          Yes   
3       Male              1     Yes         No      13          Yes   
4     Female              1     Yes         No       3          Yes   
...      ...            ...     ...        ...     ...          ...   
7262  Female              0      No         No      13          Yes   
7263    Male              0     Yes         No      22          Yes   
7264    Male              0      No         No       2          Yes   
7265    Male              0     Yes        Yes      67          Yes   
7266    Male              0     Yes        Yes      63           No   

         MultipleLines InternetService OnlineSecurity OnlineBackup  ...  \
0                   No             DSL             No          Yes  ...   
1                  Yes             DSL             No           No  ...   
2                   No     Fiber optic             No           No  ...   
3                   No     Fiber optic             No          Yes  ...   
4                   No     Fiber optic             No           No  ...   
...                ...             ...            ...          ...  ...   
7262                No             DSL            Yes           No  ...   
7263               Yes     Fiber optic             No           No  ...   
7264                No             DSL             No          Yes  ...   
7265                No             DSL            Yes           No  ...   
7266  No phone service             DSL            Yes          Yes  ...   

     TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0            Yes         Yes              No        One year              Yes   
1             No          No             Yes  Month-to-month               No   
2             No          No              No  Month-to-month              Yes   
3             No         Yes             Yes  Month-to-month              Yes   
4            Yes         Yes              No  Month-to-month              Yes   
...          ...         ...             ...             ...              ...   
7262         Yes          No              No        One year               No   
7263          No          No             Yes  Month-to-month              Yes   
7264          No          No              No  Month-to-month              Yes   
7265         Yes          No             Yes        Two year               No   
7266          No         Yes             Yes        Two year               No   

         PaymentMethod Charges_Monthly  Charges_Total  customerID Churn  
0         Mailed check           65.60          593.3  0002-ORFBO    No  
1         Mailed check           59.90          542.4  0003-MKNFE    No  
2     Electronic check           73.90         280.85  0004-TLHLJ   Yes  
3     Electronic check           98.00        1237.85  0011-IGKFF   Yes  
4         Mailed check           83.90          267.4  0013-EXCHZ   Yes  
...                ...             ...            ...         ...   ...  
7262      Mailed check           55.15          742.9  9987-LUTYD    No  
7263  Electronic check           85.10         1873.7  9992-RRAMN   Yes  
7264      Mailed check           50.30          92.75  9992-UJOEL    No  
7265      Mailed check           67.85        4627.65  9993-LHIEB    No  
7266  Electronic check           59.00         3707.6  9995-HOTOH    No  

[7267 rows x 21 columns]

In [12]:
dados_total.describe()

SeniorCitizen       tenure  Charges_Monthly
count    7267.000000  7267.000000      7267.000000
mean        0.162653    32.346498        64.720098
std         0.369074    24.571773        30.129572
min         0.000000     0.000000        18.250000
25%         0.000000     9.000000        35.425000
50%         0.000000    29.000000        70.300000
75%         0.000000    55.000000        89.875000
max         1.000000    72.000000       118.750000

In [13]:
dados_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7267 non-null   object 
 1   SeniorCitizen     7267 non-null   int64  
 2   Partner           7267 non-null   object 
 3   Dependents        7267 non-null   object 
 4   tenure            7267 non-null   int64  
 5   PhoneService      7267 non-null   object 
 6   MultipleLines     7267 non-null   object 
 7   InternetService   7267 non-null   object 
 8   OnlineSecurity    7267 non-null   object 
 9   OnlineBackup      7267 non-null   object 
 10  DeviceProtection  7267 non-null   object 
 11  TechSupport       7267 non-null   object 
 12  StreamingTV       7267 non-null   object 
 13  StreamingMovies   7267 non-null   object 
 14  Contract          7267 non-null   object 
 15  PaperlessBilling  7267 non-null   object 
 16  PaymentMethod     7267 non-null   object 


##### Como pôde ser observado, as colunas foram trazidas em outra língua, portanto cada coluna foi traduzida para facilitar o entendimento de qual dado está presente na coluna.

In [14]:
dados_total.columns = ['Genero', 'Senior','Parceiro', 'Dependentes', 'Meses_de_Contrato', 'Servico_Telefone', 'Varias_Linhas', 'Servico_Internet', 'Seguranca_Online', 'Backup_Online', 'Protecao_Dispositivo', 'Suporte_Tecnico','Streaming_TV','Streaming_Filmes', 'Contrato', 'Fatura_Online', 'Metodo_Pagamento', 'Encargos_Mensal', 'Encargos_Total','ID', 'Churn']
dados_total.head()

Genero  Senior Parceiro Dependentes  Meses_de_Contrato Servico_Telefone  \
0  Female       0      Yes         Yes                  9              Yes   
1    Male       0       No          No                  9              Yes   
2    Male       0       No          No                  4              Yes   
3    Male       1      Yes          No                 13              Yes   
4  Female       1      Yes          No                  3              Yes   

  Varias_Linhas Servico_Internet Seguranca_Online Backup_Online  ...  \
0            No              DSL               No           Yes  ...   
1           Yes              DSL               No            No  ...   
2            No      Fiber optic               No            No  ...   
3            No      Fiber optic               No           Yes  ...   
4            No      Fiber optic               No            No  ...   

  Suporte_Tecnico Streaming_TV Streaming_Filmes        Contrato Fatura_Online  \
0             Yes          Yes               No        One year           Yes   
1              No           No              Yes  Month-to-month            No   
2              No           No               No  Month-to-month           Yes   
3              No          Yes              Yes  Month-to-month           Yes   
4             Yes          Yes               No  Month-to-month           Yes   

   Metodo_Pagamento Encargos_Mensal  Encargos_Total          ID Churn  
0      Mailed check            65.6           593.3  0002-ORFBO    No  
1      Mailed check            59.9           542.4  0003-MKNFE    No  
2  Electronic check            73.9          280.85  0004-TLHLJ   Yes  
3  Electronic check            98.0         1237.85  0011-IGKFF   Yes  
4      Mailed check            83.9           267.4  0013-EXCHZ   Yes  

[5 rows x 21 columns]

### Necessário econtrar inconsistência nos dados, para isso foi realizada uma análise coluna a coluna para entender e determinar os problemas.

In [15]:
dados_total['Genero'].unique()

array(['Female', 'Male'], dtype=object)

In [16]:
dados_total['Genero'].value_counts()

Male      3675
Female    3592
Name: Genero, dtype: int64

Coluna "Genero" possui apenas os dois valores existentes, sendo bem distribuídos. Não aparenta ter nenhuma inconsistência.

In [17]:
dados_total['Senior'].unique()

array([0, 1])

In [18]:
dados_total['Senior'].value_counts()

0    6085
1    1182
Name: Senior, dtype: int64

Coluna "Senior" apresenta apenas dois valores existentes, sendo bem distribuidos. No entanto, é mais simples de entender caso os dados estivessem categorizados em "Sim" e "Não". Este "problema" será tratado mais a frente

In [19]:
dados_total['Parceiro'].unique()

array(['Yes', 'No'], dtype=object)

In [20]:
dados_total['Parceiro'].value_counts()

No     3749
Yes    3518
Name: Parceiro, dtype: int64

Coluna "Parceiro" apresenta apenas dois valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [21]:
dados_total['Dependentes'].unique()

array(['Yes', 'No'], dtype=object)

In [22]:
dados_total['Dependentes'].value_counts()

No     5086
Yes    2181
Name: Dependentes, dtype: int64

Coluna "Dependentes" apresenta apenas dois valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [23]:
dados_total['Meses_de_Contrato'].unique()

array([ 9,  4, 13,  3, 71, 63,  7, 65, 54, 72,  5, 56, 34,  1, 45, 50, 23,
       55, 26, 69, 11, 37, 49, 66, 67, 20, 43, 59, 12, 27,  2, 25, 29, 14,
       35, 64, 39, 40,  6, 30, 70, 57, 58, 16, 32, 33, 10, 21, 61, 15, 44,
       22, 24, 19, 47, 62, 46, 52,  8, 60, 48, 28, 41, 53, 68, 51, 31, 36,
       17, 18, 38, 42,  0])

In [24]:
dados_total['Meses_de_Contrato'].value_counts()

1     634
72    369
2     246
3     207
4     185
     ... 
38     60
39     59
44     54
36     50
0      11
Name: Meses_de_Contrato, Length: 73, dtype: int64

Coluna "Meses_de_Contrato" embora possua o valor "0" não aparenta nenhum inconsistência, visto que os clientes recém cadastrados podem ter um contrato com duração inferior a 1 mês.

In [25]:
dados_total['Servico_Telefone'].unique()

array(['Yes', 'No'], dtype=object)

In [26]:
dados_total['Servico_Telefone'].value_counts()

Yes    6560
No      707
Name: Servico_Telefone, dtype: int64

Coluna "Servico_Telefone" apresenta apenas dois valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [27]:
dados_total['Varias_Linhas'].unique()

array(['No', 'Yes', 'No phone service'], dtype=object)

In [28]:
dados_total['Varias_Linhas'].value_counts()

No                  3495
Yes                 3065
No phone service     707
Name: Varias_Linhas, dtype: int64

Coluna "Varias_Linhas" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [29]:
dados_total['Servico_Internet'].unique()

array(['DSL', 'Fiber optic', 'No'], dtype=object)

In [30]:
dados_total['Servico_Internet'].value_counts()

Fiber optic    3198
DSL            2488
No             1581
Name: Servico_Internet, dtype: int64

Coluna "Servico_Internet" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [31]:
dados_total['Seguranca_Online'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [32]:
dados_total['Seguranca_Online'].value_counts()

No                     3608
Yes                    2078
No internet service    1581
Name: Seguranca_Online, dtype: int64

Coluna "Seguranca_Online" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [33]:
dados_total['Backup_Online'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [34]:
dados_total['Backup_Online'].value_counts()

No                     3182
Yes                    2504
No internet service    1581
Name: Backup_Online, dtype: int64

Coluna "Backup_Online" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [35]:
dados_total['Protecao_Dispositivo'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [36]:
dados_total['Protecao_Dispositivo'].value_counts()

No                     3195
Yes                    2491
No internet service    1581
Name: Protecao_Dispositivo, dtype: int64

Coluna "Protecao_Dispositivo" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [37]:
dados_total['Suporte_Tecnico'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [38]:
dados_total['Suporte_Tecnico'].value_counts()

No                     3582
Yes                    2104
No internet service    1581
Name: Suporte_Tecnico, dtype: int64

Coluna "Suporte_Tecnico" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [39]:
dados_total['Streaming_TV'].unique()

array(['Yes', 'No', 'No internet service'], dtype=object)

In [40]:
dados_total['Streaming_TV'].value_counts()

No                     2896
Yes                    2790
No internet service    1581
Name: Streaming_TV, dtype: int64

Coluna "Streaming_TV" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [41]:
dados_total['Streaming_Filmes'].unique()

array(['No', 'Yes', 'No internet service'], dtype=object)

In [42]:
dados_total['Streaming_Filmes'].value_counts()

No                     2870
Yes                    2816
No internet service    1581
Name: Streaming_Filmes, dtype: int64

Coluna "Streaming_Filmes" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [43]:
dados_total['Contrato'].unique()

array(['One year', 'Month-to-month', 'Two year'], dtype=object)

In [44]:
dados_total['Contrato'].value_counts()

Month-to-month    4005
Two year          1743
One year          1519
Name: Contrato, dtype: int64

Coluna "Contrato" apresenta apenas três valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [45]:
dados_total['Fatura_Online'].unique()

array(['Yes', 'No'], dtype=object)

In [46]:
dados_total['Fatura_Online'].value_counts()

Yes    4311
No     2956
Name: Fatura_Online, dtype: int64

Coluna "Contrato" apresenta apenas dois valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [47]:
dados_total['Metodo_Pagamento'].unique()

array(['Mailed check', 'Electronic check', 'Credit card (automatic)',
       'Bank transfer (automatic)'], dtype=object)

In [48]:
dados_total['Metodo_Pagamento'].value_counts()

Electronic check             2445
Mailed check                 1665
Bank transfer (automatic)    1589
Credit card (automatic)      1568
Name: Metodo_Pagamento, dtype: int64

Coluna "Metodo_Pagamento" apresenta quatro valores, sendo bem distribuidos. também não parece haver nenhuma inconsistência.


In [49]:
dados_total['Encargos_Mensal'].unique()

array([65.6 , 59.9 , 73.9 , ..., 91.75, 68.8 , 67.85])

In [50]:
dados_total['Encargos_Mensal'].value_counts()

20.05     65
19.85     46
19.90     46
19.70     45
19.55     45
          ..
23.45      1
116.55     1
106.85     1
68.55      1
67.85      1
Name: Encargos_Mensal, Length: 1585, dtype: int64

In [51]:
dados_total['Encargos_Mensal'].max()

118.75

In [52]:
dados_total['Encargos_Mensal'].min()

18.25

In [53]:
dados_total.query('Encargos_Mensal == " "')

Empty DataFrame
Columns: [Genero, Senior, Parceiro, Dependentes, Meses_de_Contrato, Servico_Telefone, Varias_Linhas, Servico_Internet, Seguranca_Online, Backup_Online, Protecao_Dispositivo, Suporte_Tecnico, Streaming_TV, Streaming_Filmes, Contrato, Fatura_Online, Metodo_Pagamento, Encargos_Mensal, Encargos_Total, ID, Churn]
Index: []

[0 rows x 21 columns]

Como esta coluna possui vários valores diferentes, foi necessário realizar algumas outras análises, como foi observado o valor máximo e mínimo são valores que fazem sentido, além disso não foi encontrado algum valor faltante, portanto também parece não haver inconsistência nessa coluna.

In [54]:
dados_total['Encargos_Total'].unique()

array(['593.3', '542.4', '280.85', ..., '742.9', '4627.65', '3707.6'],
      dtype=object)

In [55]:
dados_total['Encargos_Total'].value_counts()

           11
20.2       11
19.75       9
19.55       9
19.9        9
           ..
272         1
1426.45     1
371.6       1
6786.4      1
3707.6      1
Name: Encargos_Total, Length: 6531, dtype: int64

In [56]:
dados_total['Encargos_Total'].min()

' '

In [57]:
dados_total['Encargos_Total'].max()

'999.9'

Novamente, como a variável "Encargos_Total" é uma variável quantitativa, é comum que apresente diversos valores diferentes, não sendo tão simples encontrar valores discrepantes, inconsistentes ou faltantes. Como podemos observar existem **11 valores faltantes** nos dados, podemos notá-los quando pedimos para exibir o menor valor, pela função *min()* ou pelo *value_counts()*, como feito anteriormente.

In [58]:
dados_total['ID'].unique()

array(['0002-ORFBO', '0003-MKNFE', '0004-TLHLJ', ..., '9992-UJOEL',
       '9993-LHIEB', '9995-HOTOH'], dtype=object)

In [59]:
dados_total['ID'].value_counts()

0002-ORFBO    1
6614-VBEGU    1
6637-KYRCV    1
6635-MYYYZ    1
6635-CPNUN    1
             ..
3374-TTZTK    1
3374-PZLXD    1
3374-LXDEV    1
3373-YZZYM    1
9995-HOTOH    1
Name: ID, Length: 7267, dtype: int64

A Coluna "ID" também é outra coluna que apresenta diversos valores, isso fica claro quando pensamos que cada usuário ou cliente vai ter o seu proprio ID, para se diferenciar dos demais. A coluna ID acaba não sendo muito relevante para o estudo, portanto não será feita uma análise muito detalhada dos dados para encontrar alguma inconsistência.

In [60]:
dados_total['Churn'].unique()

array(['No', 'Yes', ''], dtype=object)

In [61]:
dados_total['Churn'].value_counts()

No     5174
Yes    1869
        224
Name: Churn, dtype: int64

Por fim temos a coluna "Churn" que possui **224 Valores faltantes**, esta é uma das principais colunas, principalmente quando forem aplicados métodos de Machine Learning, portanto é necessário corrigi-los

Após Análise, Foram encontrados inconsistência em 2 colunas do DataSet, nas colunas **Churn** e **Encargos_Total**, além do "problema" da coluna **Senior** estar sendo representado de forma numérica.

### Tratamento dos dados incosistentes

#### Tratamento das Colunas "Churn" e "Encargos_Total". Em ambas colunas os valores vazios representados por " " serão substituidos por NaN a príncipio, caso seja necessário esses dados poderão ser excluídos ou substituídos por outros valores como a moda, mediana ou média da própria coluna. Esses métodos deverão ser considerados quando algum modelo de Machine Learning for aplicado. 

In [62]:
dados_total['Churn'] = dados_total['Churn'].apply(lambda x: np.nan if x == '' else x)

#### Além da inconsistencia da coluna "Encargos_Total", pode-se observar que embora esta seja uma variável quantitativa, seu tipo foi identificado pelo pandas como "Object", portanto além de substituir seus valores faltantes por NaN, seus valores serão substituídos para Float

In [63]:
dados_total['Encargos_Total'] = dados_total['Encargos_Total'].apply(lambda x: np.nan if x == ' ' else float(x))

In [64]:
dic_senior = {0: 'Não', 1: 'Sim'}
dados_total['Senior'].replace(dic_senior, inplace = True)

Todas as correções foram realizadas, o código a seguir mostra que os valores faltantes já estão sendo reconhecidos pelo Python, sendo mais fácil de tratá-los posteriormente.

In [65]:
dados_total.isnull().sum()

Genero                    0
Senior                    0
Parceiro                  0
Dependentes               0
Meses_de_Contrato         0
Servico_Telefone          0
Varias_Linhas             0
Servico_Internet          0
Seguranca_Online          0
Backup_Online             0
Protecao_Dispositivo      0
Suporte_Tecnico           0
Streaming_TV              0
Streaming_Filmes          0
Contrato                  0
Fatura_Online             0
Metodo_Pagamento          0
Encargos_Mensal           0
Encargos_Total           11
ID                        0
Churn                   224
dtype: int64

In [66]:
dados_total.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Genero                7267 non-null   object 
 1   Senior                7267 non-null   object 
 2   Parceiro              7267 non-null   object 
 3   Dependentes           7267 non-null   object 
 4   Meses_de_Contrato     7267 non-null   int64  
 5   Servico_Telefone      7267 non-null   object 
 6   Varias_Linhas         7267 non-null   object 
 7   Servico_Internet      7267 non-null   object 
 8   Seguranca_Online      7267 non-null   object 
 9   Backup_Online         7267 non-null   object 
 10  Protecao_Dispositivo  7267 non-null   object 
 11  Suporte_Tecnico       7267 non-null   object 
 12  Streaming_TV          7267 non-null   object 
 13  Streaming_Filmes      7267 non-null   object 
 14  Contrato              7267 non-null   object 
 15  Fatura_Online        

In [67]:
dados_total.head()

Genero Senior Parceiro Dependentes  Meses_de_Contrato Servico_Telefone  \
0  Female    Não      Yes         Yes                  9              Yes   
1    Male    Não       No          No                  9              Yes   
2    Male    Não       No          No                  4              Yes   
3    Male    Sim      Yes          No                 13              Yes   
4  Female    Sim      Yes          No                  3              Yes   

  Varias_Linhas Servico_Internet Seguranca_Online Backup_Online  ...  \
0            No              DSL               No           Yes  ...   
1           Yes              DSL               No            No  ...   
2            No      Fiber optic               No            No  ...   
3            No      Fiber optic               No           Yes  ...   
4            No      Fiber optic               No            No  ...   

  Suporte_Tecnico Streaming_TV Streaming_Filmes        Contrato Fatura_Online  \
0             Yes          Yes               No        One year           Yes   
1              No           No              Yes  Month-to-month            No   
2              No           No               No  Month-to-month           Yes   
3              No          Yes              Yes  Month-to-month           Yes   
4             Yes          Yes               No  Month-to-month           Yes   

   Metodo_Pagamento Encargos_Mensal  Encargos_Total          ID Churn  
0      Mailed check            65.6          593.30  0002-ORFBO    No  
1      Mailed check            59.9          542.40  0003-MKNFE    No  
2  Electronic check            73.9          280.85  0004-TLHLJ   Yes  
3  Electronic check            98.0         1237.85  0011-IGKFF   Yes  
4      Mailed check            83.9          267.40  0013-EXCHZ   Yes  

[5 rows x 21 columns]

Podemos ver que agora aparecem os mesmos valores de dados faltantes que os encontrados anteriormente, além da variável "Encargos_total" agora ser do tipo Float. E como citado anteriormente, agora os dados da coluna "Senior" agora são do tipo categóricos "Sim" e "Não"

### Além de traduzir todas as colunas, agora será traduzidos os seus valores, para facilitar ainda mais o entendimento.

In [68]:
dic_sim_nao = {'Yes': 'Sim', 'No': 'Não'}

colunas = ['Churn', 'Parceiro', 'Dependentes', 'Servico_Telefone', 
           'Varias_Linhas', 'Seguranca_Online', 'Backup_Online', 'Protecao_Dispositivo', 
           'Suporte_Tecnico', 'Streaming_TV', 'Streaming_Filmes', 'Fatura_Online']
for x in colunas:
  dados_total[x].replace(dic_sim_nao, inplace = True)


In [69]:
dic_homem_mulher = {'Male': 'Homem', 'Female': 'Mulher'}

dados_total['Genero'].replace(dic_homem_mulher, inplace = True)

In [70]:
dados_total['Servico_Telefone'].replace('No phone service', 'Sem serviço de Telefone', inplace = True)
dados_total['Varias_Linhas'].replace('No phone service', 'Sem serviço de Telefone', inplace = True)

In [71]:
dic_sem_internet = {'No internet service': 'Sem serviço de internet'}
colunas = ['Seguranca_Online', 'Backup_Online', 'Protecao_Dispositivo','Suporte_Tecnico', 'Streaming_TV', 'Streaming_Filmes']

for x in colunas: 
  dados_total[x].replace(dic_sem_internet, inplace = True)

In [72]:
dados_total['Servico_Internet'].replace('Fiber optic', 'Fibra Óptica', inplace = True)

In [73]:
dic_contrato = {'Month-to-month': 'Mensal', 'Two year': 'Dois Anos', 'One year': 'Um Ano'}

dados_total['Contrato'].replace(dic_contrato, inplace = True)

In [74]:
dic_pagamento = {'Mailed check': 'Cheque', 'Electronic check': 'Cheque Eletrônico', 'Credit card (automatic)': 'Cartão de crédito', 'Bank transfer (automatic)': 'Transferência Bancária'}

dados_total['Metodo_Pagamento'].replace(dic_pagamento, inplace = True)

### Criando a coluna de contas diárias

In [75]:
dados_total['Encargos_Diarios'] = round(dados_total['Encargos_Mensal'] / 30, 2)

In [76]:
dados_total

Genero Senior Parceiro Dependentes  Meses_de_Contrato Servico_Telefone  \
0     Mulher    Não      Sim         Sim                  9              Sim   
1      Homem    Não      Não         Não                  9              Sim   
2      Homem    Não      Não         Não                  4              Sim   
3      Homem    Sim      Sim         Não                 13              Sim   
4     Mulher    Sim      Sim         Não                  3              Sim   
...      ...    ...      ...         ...                ...              ...   
7262  Mulher    Não      Não         Não                 13              Sim   
7263   Homem    Não      Sim         Não                 22              Sim   
7264   Homem    Não      Não         Não                  2              Sim   
7265   Homem    Não      Sim         Sim                 67              Sim   
7266   Homem    Não      Sim         Sim                 63              Não   

                Varias_Linhas Servico_Internet Seguranca_Online Backup_Online  \
0                         Não              DSL              Não           Sim   
1                         Sim              DSL              Não           Não   
2                         Não     Fibra Óptica              Não           Não   
3                         Não     Fibra Óptica              Não           Sim   
4                         Não     Fibra Óptica              Não           Não   
...                       ...              ...              ...           ...   
7262                      Não              DSL              Sim           Não   
7263                      Sim     Fibra Óptica              Não           Não   
7264                      Não              DSL              Não           Sim   
7265                      Não              DSL              Sim           Não   
7266  Sem serviço de Telefone              DSL              Sim           Sim   

      ... Streaming_TV Streaming_Filmes   Contrato Fatura_Online  \
0     ...          Sim              Não     Um Ano           Sim   
1     ...          Não              Sim     Mensal           Não   
2     ...          Não              Não     Mensal           Sim   
3     ...          Sim              Sim     Mensal           Sim   
4     ...          Sim              Não     Mensal           Sim   
...   ...          ...              ...        ...           ...   
7262  ...          Não              Não     Um Ano           Não   
7263  ...          Não              Sim     Mensal           Sim   
7264  ...          Não              Não     Mensal           Sim   
7265  ...          Não              Sim  Dois Anos           Não   
7266  ...          Sim              Sim  Dois Anos           Não   

       Metodo_Pagamento Encargos_Mensal Encargos_Total          ID  Churn  \
0                Cheque           65.60         593.30  0002-ORFBO    Não   
1                Cheque           59.90         542.40  0003-MKNFE    Não   
2     Cheque Eletrônico           73.90         280.85  0004-TLHLJ    Sim   
3     Cheque Eletrônico           98.00        1237.85  0011-IGKFF    Sim   
4                Cheque           83.90         267.40  0013-EXCHZ    Sim   
...                 ...             ...            ...         ...    ...   
7262             Cheque           55.15         742.90  9987-LUTYD    Não   
7263  Cheque Eletrônico           85.10        1873.70  9992-RRAMN    Sim   
7264             Cheque           50.30          92.75  9992-UJOEL    Não   
7265             Cheque           67.85        4627.65  9993-LHIEB    Não   
7266  Cheque Eletrônico           59.00        3707.60  9995-HOTOH    Não   

     Encargos_Diarios  
0                2.19  
1                2.00  
2                2.46  
3                3.27  
4                2.80  
...               ...  
7262             1.84  
7263             2.84  
7264             1.68  
7265             2.26  
7266             1.97  

[7267 rows x 22 columns]

## Criando uma cópia do dataset, nesta cópia todos os valores faltantes serão deletados.

In [77]:
dados_total_limpo = dados_total.copy()

In [78]:
dados_total_limpo.head()

Genero Senior Parceiro Dependentes  Meses_de_Contrato Servico_Telefone  \
0  Mulher    Não      Sim         Sim                  9              Sim   
1   Homem    Não      Não         Não                  9              Sim   
2   Homem    Não      Não         Não                  4              Sim   
3   Homem    Sim      Sim         Não                 13              Sim   
4  Mulher    Sim      Sim         Não                  3              Sim   

  Varias_Linhas Servico_Internet Seguranca_Online Backup_Online  ...  \
0           Não              DSL              Não           Sim  ...   
1           Sim              DSL              Não           Não  ...   
2           Não     Fibra Óptica              Não           Não  ...   
3           Não     Fibra Óptica              Não           Sim  ...   
4           Não     Fibra Óptica              Não           Não  ...   

  Streaming_TV Streaming_Filmes Contrato Fatura_Online   Metodo_Pagamento  \
0          Sim              Não   Um Ano           Sim             Cheque   
1          Não              Sim   Mensal           Não             Cheque   
2          Não              Não   Mensal           Sim  Cheque Eletrônico   
3          Sim              Sim   Mensal           Sim  Cheque Eletrônico   
4          Sim              Não   Mensal           Sim             Cheque   

  Encargos_Mensal Encargos_Total          ID  Churn Encargos_Diarios  
0            65.6         593.30  0002-ORFBO    Não             2.19  
1            59.9         542.40  0003-MKNFE    Não             2.00  
2            73.9         280.85  0004-TLHLJ    Sim             2.46  
3            98.0        1237.85  0011-IGKFF    Sim             3.27  
4            83.9         267.40  0013-EXCHZ    Sim             2.80  

[5 rows x 22 columns]

In [79]:
dados_total_limpo.dropna(inplace = True)

In [80]:
dados_total_limpo

Genero Senior Parceiro Dependentes  Meses_de_Contrato Servico_Telefone  \
0     Mulher    Não      Sim         Sim                  9              Sim   
1      Homem    Não      Não         Não                  9              Sim   
2      Homem    Não      Não         Não                  4              Sim   
3      Homem    Sim      Sim         Não                 13              Sim   
4     Mulher    Sim      Sim         Não                  3              Sim   
...      ...    ...      ...         ...                ...              ...   
7262  Mulher    Não      Não         Não                 13              Sim   
7263   Homem    Não      Sim         Não                 22              Sim   
7264   Homem    Não      Não         Não                  2              Sim   
7265   Homem    Não      Sim         Sim                 67              Sim   
7266   Homem    Não      Sim         Sim                 63              Não   

                Varias_Linhas Servico_Internet Seguranca_Online Backup_Online  \
0                         Não              DSL              Não           Sim   
1                         Sim              DSL              Não           Não   
2                         Não     Fibra Óptica              Não           Não   
3                         Não     Fibra Óptica              Não           Sim   
4                         Não     Fibra Óptica              Não           Não   
...                       ...              ...              ...           ...   
7262                      Não              DSL              Sim           Não   
7263                      Sim     Fibra Óptica              Não           Não   
7264                      Não              DSL              Não           Sim   
7265                      Não              DSL              Sim           Não   
7266  Sem serviço de Telefone              DSL              Sim           Sim   

      ... Streaming_TV Streaming_Filmes   Contrato Fatura_Online  \
0     ...          Sim              Não     Um Ano           Sim   
1     ...          Não              Sim     Mensal           Não   
2     ...          Não              Não     Mensal           Sim   
3     ...          Sim              Sim     Mensal           Sim   
4     ...          Sim              Não     Mensal           Sim   
...   ...          ...              ...        ...           ...   
7262  ...          Não              Não     Um Ano           Não   
7263  ...          Não              Sim     Mensal           Sim   
7264  ...          Não              Não     Mensal           Sim   
7265  ...          Não              Sim  Dois Anos           Não   
7266  ...          Sim              Sim  Dois Anos           Não   

       Metodo_Pagamento Encargos_Mensal Encargos_Total          ID  Churn  \
0                Cheque           65.60         593.30  0002-ORFBO    Não   
1                Cheque           59.90         542.40  0003-MKNFE    Não   
2     Cheque Eletrônico           73.90         280.85  0004-TLHLJ    Sim   
3     Cheque Eletrônico           98.00        1237.85  0011-IGKFF    Sim   
4                Cheque           83.90         267.40  0013-EXCHZ    Sim   
...                 ...             ...            ...         ...    ...   
7262             Cheque           55.15         742.90  9987-LUTYD    Não   
7263  Cheque Eletrônico           85.10        1873.70  9992-RRAMN    Sim   
7264             Cheque           50.30          92.75  9992-UJOEL    Não   
7265             Cheque           67.85        4627.65  9993-LHIEB    Não   
7266  Cheque Eletrônico           59.00        3707.60  9995-HOTOH    Não   

     Encargos_Diarios  
0                2.19  
1                2.00  
2                2.46  
3                3.27  
4                2.80  
...               ...  
7262             1.84  
7263             2.84  
7264             1.68  
7265             2.26  
7266             1.97  

[7032 rows x 22 columns]

In [82]:
dados_total_limpo.isnull().sum()

Genero                  0
Senior                  0
Parceiro                0
Dependentes             0
Meses_de_Contrato       0
Servico_Telefone        0
Varias_Linhas           0
Servico_Internet        0
Seguranca_Online        0
Backup_Online           0
Protecao_Dispositivo    0
Suporte_Tecnico         0
Streaming_TV            0
Streaming_Filmes        0
Contrato                0
Fatura_Online           0
Metodo_Pagamento        0
Encargos_Mensal         0
Encargos_Total          0
ID                      0
Churn                   0
Encargos_Diarios        0
dtype: int64

## Exportando os dois Datasets criados


In [81]:
dados_total.to_csv('dados_clientes_alura.csv', index = False)
dados_total_limpo.to_csv('dados_clientes_limpo_alura.csv', index = False)